tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Reporter.cs

```diff
@@ -62,68 +67,105 @@ private static void LogAddressIfDebugEnabled(IDictionary<string, object> args)
-        {
-            _localRootSpan.SetTag(Tags.ActorIp, clientIp);
-        }
+            LogMatchesIfDebugEnabled(result.Data, blocked);
 
-        if (_localRootSpan.Context.TraceContext is { Origin: null } traceContext)
-        {
-            _localRootSpan.SetTag(Tags.Origin, "appsec");
-            traceContext.Origin = "appsec";
-        }
+            _localRootSpan.SetTag(Tags.AppSecJson, "{\"triggers\":" + result.Data + "}");
+            var clientIp = _localRootSpan.GetTag(Tags.HttpClientIp);
+            if (!string.IsNullOrEmpty(clientIp))
+            {
+                _localRootSpan.SetTag(Tags.ActorIp, clientIp);
+            }
+
+            if (_localRootSpan.Context.TraceContext is { Origin: null } traceContext)
+            {
+                _localRootSpan.SetTag(Tags.Origin, "appsec");
+                traceContext.Origin = "appsec";
+            }
 
-        _localRootSpan.SetTag(Tags.AppSecRuleFileVersion, _security.WafRuleFileVersion);
-        _localRootSpan.SetMetric(Metrics.AppSecWafDuration, aggregatedTotalRuntime);
-        _localRootSpan.SetMetric(Metrics.AppSecWafAndBindingsDuration, aggregatedTotalRuntimeWithBindings);
-        var headers = _httpTransport.GetRequestHeaders();
-        AddHeaderTags(_localRootSpan, headers, RequestHeaders, SpanContextPropagator.HttpRequestHeadersTagPrefix);
+            _localRootSpan.SetTag(Tags.AppSecRuleFileVersion, _security.WafRuleFileVersion);
+            _localRootSpan.SetMetric(Metrics.AppSecWafDuration, result.AggregatedTotalRuntime);
+            _localRootSpan.SetMetric(Metrics.AppSecWafAndBindingsDuration, result.AggregatedTotalRuntimeWithBindings);
+            var headers = _httpTransport.GetRequestHeaders();
+            AddHeaderTags(_localRootSpan, headers, RequestHeaders, SpanContextPropagator.HttpRequestHeadersTagPrefix);
 
-        if (status is not null)
+            if (status is not null)
+            {
+                _localRootSpan.SetHttpStatusCode(status.Value, isServer: true, Tracer.Instance.Settings);
+            }
+        }
+
+        if (result.ShouldReportSchema)
         {
-            _localRootSpan.SetHttpStatusCode(status.Value, isServer: true, Tracer.Instance.Settings);
+            foreach (var derivative in result.Derivatives)
+            {
+                var serializeObject = JsonConvert.SerializeObject(derivative.Value);
+                var bytes = Encoding.UTF8.GetBytes(serializeObject);
+                if (bytes.Length <= MaxApiSecurityTagValueLength)
+                {
+                    var memoryStream = new MemoryStream();
+                    using (var gZipStream = new GZipStream(memoryStream, CompressionMode.Compress))
+                    {
+                        gZipStream.Write(bytes, 0, bytes.Length);
+                    }
+
+                    if (memoryStream.TryGetBuffer(out var bytesResult))
+                    {
+                        var gzipBase64 = Convert.ToBase64String(bytesResult.Array!);
```

Comment:
When using `ArraySegment<T>` you _must_ use the `Offset` and `Count` properties. The underlying `Array` value potentially (almost certainly) contains _extra_ data, that you _don't_ want to encode.

```suggestion
                        var gzipBase64 = Convert.ToBase64String(bytesResult.Array!, bytesResult.Offset, bytesResult.Count);
```